main.js

```diff
@@ -170,98 +343,361 @@ function render2D(foldObj) {
+                }
+            }
+            return outArray;
+        }
+
+        /**
+         * Helper function to cancatenate two arrays even if they have nested arrays inside
+         * @param {Array} array1 The array to be concatenated to
+         * @param {Array} array2 The array to concatenate onto array1
+         * @returns {Array} The result of deep concatenation
+         */
+        function deepArrayConcat(array1, array2) {
+            // TODO: handle in more graceful manner
+            if (!array2) {
+                alert("Something went wrong with coordinate arrays");
+            }
+            if (array2.some(el => Array.isArray(el))) {
+                array2.forEach(element => {
+                    if (Array.isArray(element)) {
+                        array1 = deepArrayConcat(array1, element);
+                    } else {
+                        array1 = array1.concat(element);
+                    }
+                });
+                return array1;
+            } else {
+                array1 = array1.concat(array2);
+                return array1;
+            }
+        }
+
+        /**
+         * Function to split an array of indexed vertices as faces into triangles if they aren't already
+         * @param {Array} array A list of faces made by referencing indexed vertices
+         * @returns {Array} The original array modified to have many triangles instead of polygons
+         */
+        function polygonToTri(array) {
+            let outArray = [];
+            for (let i = 0; i < array.length; i++) {
+                let len = array[i].length;
+                if (len === 3) {
+                    outArray.push(array[i]);
+                } else if (len === 4) { // unsure about higher polygons so only handle rect
+                    // get the last three elements, then get the first three elements
+                    outArray.push(array[i].slice(0, 3));
+                    /*
+                     * since the face has its vertices listed in a consistent counterclockwise or
+                     * clockwise order, we don't need the first three then last three, we need
+                     * the first three, then the triangle opposite that is formed by the last two
+                     * vertices and the first one
+                     * this does that by moving the first element
+                     */
+                    let shifted = array[i];
+                    shifted.push(array[i].shift())
+                    outArray.push(shifted.slice(-3));
+                } else if (len < 3 || len > 4) {
+                    // TODO: replace with better alert when frontend is more permanent
+                    alert("Polygon rendering encountered unexpected shapes");
+                }
+            }
+            return outArray;
+        }
+
+        /**
+         * Helper function to convert from XYZ (FOLD) to XZY (THREE.js) and possibly back as well. Used
```

Comment:
Are you sure that THREE uses an XZY format? That would surprise me quite a bit. I think this might be an artifact of your camera positioning.

Reply:
Not really an XZY format? I think this may be a little unclear. But I do think that FOLD treats the Z axis as upwards, and THREE treats the Y axis as upwards. Since I was working with them all the same later, this function swaps the last two.
Initially the camera is set to some value along the Z axis, and it looks by default towards +Z. That means up is Y. Fold uses (x, y, z) for the vertices location, but since it is math based I assumed that these coordinates use Z as the up/down axis.

Reply:
I think I assumed this just because folding from a flat sheet like the square-twist example I imagined that "flat" meant on the plane with the two axes neither being the "up" axis. But looking at the fold viewer example from FOLD repo has the default view to indeed look like what I saw when I was using the same "up" axis and not converting between FOLD and THREE, so I will look into it more I guess.